internal/authproxy/serve.go

```diff
@@ -32,76 +37,90 @@ import (
 // The configuration options for the authentication proxy used for anonymous users.
 // The fields below can be passed as arguments i.e. --token=some-very-complicated-random-value
 // or as a yaml config file.
-const RemoteFlag = "remote"
-const PortFlag = "port"
-const TokenFlag = "token"
-const CookieKeyFlag = "cookie_key"
-const VerboseFlag = "verbose"
+const remoteFlag = "remote"
+const portFlag = "port"
+const tokenFlag = "token"
+const cookieKeyFlag = "cookie_key"
+const verboseFlag = "verbose"
+const configFlag = "config"
+
+var remote string
+var port int
+var token string
+var cookieKey string
+var verbose bool
+var config string
+
+const prefix = "authproxy"
 
 func Command() (*cobra.Command, error) {
 	var serveCmd = &cobra.Command{
-		Use:    "serve",
-		Short:  "Run the proxy",
-		Run:    serve,
-		PreRun: initConfig,
+		Use:     "serve",
+		Short:   "Run the proxy",
+		Run:     serve,
+		PreRunE: loadConfig,
 	}
-	serveCmd.PersistentFlags().String(RemoteFlag, "", "remote URL to proxy to")
-	err := serveCmd.MarkPersistentFlagRequired(RemoteFlag)
+
+	serveCmd.PersistentFlags().StringVar(&remote, remoteFlag, "", "remote URL to proxy to")
+	err := serveCmd.MarkPersistentFlagRequired(remoteFlag)
 	if err != nil {
 		return nil, err
 	}
-	err = viper.BindPFlag(RemoteFlag, serveCmd.PersistentFlags().Lookup(RemoteFlag))
+	err = viper.BindPFlag(prefix+"."+remoteFlag, serveCmd.PersistentFlags().Lookup(remoteFlag))
 	if err != nil {
 		return nil, err
 	}
-	err = viper.BindEnv(RemoteFlag)
+	err = viper.BindEnv(prefix+"."+remoteFlag, strings.ToUpper(prefix+"_"+remoteFlag))
 	if err != nil {
 		return nil, err
 	}
 
-	serveCmd.PersistentFlags().Int(PortFlag, 65535, "port on which the proxy will listen")
-	err = viper.BindPFlag(PortFlag, serveCmd.PersistentFlags().Lookup(PortFlag))
+	serveCmd.PersistentFlags().IntVar(&port, portFlag, 65535, "port on which the proxy will listen")
+	err = viper.BindPFlag(prefix+"."+portFlag, serveCmd.PersistentFlags().Lookup(portFlag))
 	if err != nil {
 		return nil, err
 	}
-	err = viper.BindEnv(PortFlag)
+	err = viper.BindEnv(prefix+"."+portFlag, strings.ToUpper(prefix+"_"+portFlag))
```

Comment:
I just realized, aren't your manually doing what `Viper.SetEnvPrefix("authproxy")` and `viper.AutomaticEnv()` do ?

See [here](https://github.com/spf13/viper?tab=readme-ov-file#env-example) for what I had in mind.

Reply:
This was such a pain to debug. And for some reason I could not get it to work with flags, the files and env variables when `viper.SetEnvPrefix` was used. 

The reason for this is that when you use AutomaicEnv and SetEnvPrefix then it assumes that the name of the variable in viper equals the name of the env variable. But this is not the case. We want the names in viper to have a prefix like `section.name` so that they can be loaded from a yaml file like 

```yaml
section:
  name: secret-value
```

But when you use automatic env then it puts the name of the variable in viper under just `name`. It removes the prefix. So then flags and the secret file get saved in a different name in viper compared to env secrets. This was the only way I could work out to make sure that we have sections in the yaml file for different things - not just one flat dictionary and to have the secrets and the flags all work and populate the same place in viper.

Reply:
In the future I will remove the options to configure the sidecar container through so many things. We dont have to accept cli args, config file and env vars all at the same time and try to merge all of them.

Reply:
Ok. all good then, thanks.